halo/attest/keeper/keeper_test.go

```diff
@@ -0,0 +1,180 @@
+package keeper_test
+
+import (
+	"testing"
+
+	"github.com/omni-network/omni/halo/attest/keeper"
+	"github.com/omni-network/omni/halo/attest/types"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
+	"github.com/stretchr/testify/require"
+)
+
+func TestKeeper_Add(t *testing.T) {
+	t.Parallel()
+
+	// cmp transformation options to ignore private fields of proto generated types.
+	var (
+		atteCmpOpts = cmp.Options{cmpopts.IgnoreUnexported(keeper.Attestation{})}
+		sigsCmpOpts = cmp.Options{cmpopts.IgnoreUnexported(keeper.Signature{})}
+	)
+
+	type args struct {
+		msg *types.MsgAddVotes
+	}
+	type want struct {
+		atts []*keeper.Attestation
+		sigs []*keeper.Signature
+	}
+
+	tests := []struct {
+		name          string
+		expectations  []func(sdk.Context, mocks)                              // These functions set expectations in the various mocked dependencies.
+		prerequisites []func(t *testing.T, k *keeper.Keeper, ctx sdk.Context) // These functions modify the keeper before calling its Add method.
+		args          args
+		want          want
+		wantErr       bool
+	}{
+		{
+			name: "single_vote",
+			args: args{
+				msg: defaultMsg().Msg(),
+			},
+			want: want{
+				atts: []*keeper.Attestation{
+					{Id: 1, BlockRoot: blockRoot, ChainId: 1, Hash: blockHashes[0].Bytes(), Height: 500, Status: int32(keeper.Status_Pending)},
+				},
+				sigs: []*keeper.Signature{
+					{Id: 1, AttId: 1, Signature: val1.Bytes(), ValidatorAddress: val1.Address},
+					{Id: 2, AttId: 1, Signature: val2.Bytes(), ValidatorAddress: val2.Address},
+				},
+			},
+		},
+		{
+			name: "two_votes_diff_blocks",
+			args: args{
+				msg: defaultMsg().
+					WithAppendVotes(
+						new(AggVoteBuilder).Default().WithBlockHeader(1, 501, blockHashes[1]).WithSignatures(sigsTuples(val1, val3)...).Vote(),
```

Comment:
cant we have a `defaultAggVote()` as well? I see a lot of `new(AggVoteBuilder).Default()` duplication